Update healthz documentation. (#106)
openconfig/gnoi: healthz/README.md

# gNOI `Healthz` Streaming RPC Design
**Contributors**: hines@google.com, robjs@google.com, bneville@arista.com, roman.dodin@nokia.com
**Last Updated**: 2022-11-02

## Background

>  * [gNOI Repository](https://github.com/openconfig/gnoi)
>  * [gNOI `Healthz` service](https://github.com/openconfig/gnoi/tree/master/healthz)

Reviewer: linting: remove extra space in the unordered list
```suggestion
* [gNOI Repository](https://github.com/openconfig/gnoi)
* [gNOI `Healthz` service](https://github.com/openconfig/gnoi/tree/master/healthz)
```

The ability to also include "implementation specific" details about the health
of the system or component is very standard for network devices in the form for
"show tech", "diag" or "debug" commands. These are very useful for getting
diagnostics about the system state to the implementers. Healthz exposes
these interfaces as queriable endpoints to allow operators to validate "health"
on components and if "unhealthy" gather implementor specific data to help
triage or reproduce issues.

## Design Background

`Healthz` provides a means by which a user may initiate checks of health
(through the `Check` RPC) or a system may report the results of a check that it

> has initiated of its own accord. 

Reviewer: re "a system may report the results of a check that it has initiated of its own accord.":

1. when a system boots, should the relevant component manager perform initial health checks and thus generate ComponentStatus for components? 
2. When a certain component changes its health status from one state to another (think a tranceiver goes from healthy to unhealthy), should the relevant component manager record a ComponentStatus event with a distinct ID automatically?

Reviewer 2: i would say it should look like this:

system boots
gnmi tree is initialized for things which have not run a check yet - they start in the unspecfied state.
During initialization of the subcomponents the system should perform initial health checks to put the components into a known state 
if the component is unhealthy - generate an event
if the components is healthy - set state and move on

2. yes if something goes unhealthly there should be an event - but how much work is done to "triage" that event is left to the vendor.

> Following a health check occuring, a caller uses the `List` RPC to discover the

Reviewer: Does it make sense to specify `Get` RPC as an option in the workflow described in this sentence?
```suggestion
Following a health check occurring, a caller uses the `List` or `Get` RPC to discover the
```

> artifacts that are assoicated with a component and its corresponding

Reviewer: ```suggestion
artifacts that are associated with a component and its corresponding
```

> artifacts that are assoicated with a component and its corresponding

Reviewer: It seems confusing to use the `artifact` term when we describe the `ComponentStatus`, as the artifact is a message encompassed in the status.
Later down the section we say that we acknowledge artifacts:

```
Once retrieved each artifact can be 'acknowledged' by the client of the RPC.
```

which confuses me, as we ack the `ComponentStatus` (which is also referred as `event` sometimes in this doc).

Shall we rather say `ComponentStatus` when describing the set of health status messages returned by `Get`/`List` RPCs?

Author: I agree -- I've tried to standardise that "event" means a `ComponentStatus`. I think that this is clearer, since `ComponentStatus` itself seems obsure from a human understanding. PTAL (will push this change after sending this comment and addressing the others).

subcomponents. Each artifact reflects an entity that is required to debug or
further root cause the fault that occurs with it.

The `Artifact` RPC is used to retrieve specific artifacts that are listed by

> the target system in the `List` RPC. Once retrieved each artifact can be

Reviewer: ```suggestion
the target system in the `List` or `Get` RPC. Once retrieved each artifact can be
```

'acknowledged' by the client of the RPC. Acknowledged artifacts are no longer
returned in a list of the corresponding artifacts, and a device may use this
status as a hint to allow garbage collection of artifacts that are no longer
relevant. The device itself is responsible for garbage collection any may,

> if necessary, garbage collect artifacts that are not yet acknowledged.

Reviewer 2: they should not be cleared just because the gnoi server restarted

> if necessary, garbage collect artifacts that are not yet acknowledged.

Author: Added a clarification -- I think generally these need to persist across component restarts too, since healthz information that corresponds to something like a rebooting linecard should not be removed as the LC reboots IMHO.

> if necessary, garbage collect artifacts that are not yet acknowledged.

Reviewer: so essentially we have to persist the events until they are garbage collected/acked.

of the component's health is required. The following paths (defined in the
`openconfig-platform` model) are used:

```
component/healthz/state/status          (enumerated value)
component/healthz/state/last-unhealthy  (timestamp)
component/healthz/state/unhealthy-count (int64)
```

### Choice of Streaming for Artifacts

[Background comment](https://github.com/openconfig/gnoi/pull/65#issuecomment-1090547841)

Originally, the gNOI Healthz endpoint defined single transactional API for
getting the "healthz" status of a component.  As discussed in the above issue,
due to large data volume or to an interactive debug output there is a
requirement that it support the ability to stream a status from the device.  

To ensure that this can be done using a secure channel with the relevant
ACL and encryption, this streaming is implemented within the `Healthz` service.
This approach further allows the concept of a historical set of collected
articacts to be reported by the device.

## Architecture

### Use of gNMI paths

Where a gNMI path (`gnoi.types.Path`) is used in `Healthz`, the path specified
should be the complete path to a specific component - i.e.,

> `/components/component[name=FOO]`.

Author: BGP itself would not necessarily be exposed as a component, but it is valid to have software processes in the `/components` tree. Based on this, I don't _think_ that there is a need to restrict this if the implementation supports a health check on software processes.

> `/components/component[name=FOO]`.

Reviewer: @robshakir okay, my reading here is that it is not prescribed by healthz (or Openconfig) to expose software services under `/components`, moreover, the schema for `/components` container hints that it is meant to be used for HW components as it has sub-containers like chassis, CPU, etc.

Therefore, it seems okay to first focus on the hardware parts in the early days of healthz and approach SW components later

The ability to also include "implementation specific" details about the health
of the system or component is very standard for network devices in the form for
"show tech", "diag" or "debug" commands. These are very useful for getting
diagnostics about the system state to the implementers. Healthz exposes
these interfaces as queriable endpoints to allow operators to validate "health"
on components and if "unhealthy" gather implementor specific data to help
triage or reproduce issues.

## Design Background

`Healthz` provides a means by which a user may initiate checks of health
(through the `Check` RPC) or a system may report the results of a check that it
has initiated of its own accord. 

Following a health check occurring, a caller uses the `List` or `Get` RPC to discover the
health "events" (expressed as a `ComponentStatus` message) that are assoicated

> with a component and its corresponding subcomponents. Each event reflects an

Reviewer: ```suggestion
with a component and its corresponding subcomponents. Each event reflects a
```

collection of data that is required to debug or further root cause the fault
that occurs with an entity in the system.

The `Artifact` RPC is used to retrieve specific artifacts that are listed by
the target system in the `List` or `Get` RPC. Once retrieved each artifact can be
'acknowledged' by the client of the RPC. Acknowledged artifacts are no longer
returned in a list of the corresponding artifacts, and a device may use this

> status as a hint to allow garbage collection of artifacts that are no longer

Author: Done.

> status as a hint to allow garbage collection of artifacts that are no longer

Reviewer: I don't see changes to this paragraph in [79095d3](https://github.com/openconfig/gnoi/pull/106/commits/79095d3f6a11d9db4e60e4c282144c298a3bd563). But I am fine with merging this as is to not let this one stall

> status as a hint to allow garbage collection of artifacts that are no longer

Author: Sorry - just waiting to push it whilst working on the other item. Will have it this afternoon.

with a specific component and any of its corresponding subcomponents. In contrast to
`List` the `Get` RPC should return only the latest event (i.e.,
`ComponentStatus` message).

The `GetResponse` returned includes a `ComponentStatus` message which
corresponds to the latest health event for the  component itself and each subcomponent
being reported on. Thus, multiple `ComponentStatus` messages may be reported for
a single component. Each message includes a set of `ArtifactHeader` messages
that correspond to the health event -- and provide identifiers and types for the
artifacts that can be returned by the system.

All artifacts that are listed within the same `ComponentStatus` message are
expected to have a shared acknowledgement state, and expiry time.

### Healthz.List()

```protobuf
rpc List(ListRequest) returns (ListResponse) {}
```

`List` returns all events that are associated with a particular component. In
contrast to `Get`, it returns all events that the target has for the device. By
default, events that are acknowledged are not returned.

Similarly to `Get`, a series of `ComponentStatus` messages are returned with

> the same semantics as are described above.

Author: Made a slightly wider change.

```

`Artifact` allows a user to retrieve a specific artifact that is related with
an event that has occurred. Since these artifacts may be large, the `Artifact`
RPC is implemented as a server-side streaming RPC. Use of the `Artifact` RPC
ensures that the resources to send these potentially large artifacts only
when explicitly requested by the client.

#### Healthz.Check()

```protobuf
rpc Check(CheckRequest) returns (CheckResponse) {}
```

`Check` allows a client to execute a set of "validations" against the specified
component. The component, as with other operations, is specified in terms of
the gNMI path.  

The result of the `Check` produces a healthz `ComponentStatus` message which
will contain a list of the generated artifacts used in the validation process.

A call to the `Check` endpoint will allow for "very expensive" debugging
commands such as causing the device to pause and snapshot its system database
or to examine the state of a protocol.  These commands will likely be
considered "service impacting" and should have a clear security ACL restricting
their use during normal operations of the device.

## User Experience

> #### A BGP routing process goes unhealthy due to a crash.

Reviewer: next level heading should be level 3 `###`
```suggestion
### A BGP routing process goes unhealthy due to a crash.
```

> ##### Expected Action

Reviewer: ```suggestion
#### Expected Action
```

Write the core dump to a file system.

A call to `Healthz.Get` should respond with the current state of the component
and provide feedback that there is an artifact available to be requested.

> A Healthz call should stream the core dump over the service back to the caller.

Reviewer: ```suggestion
A `Healthz.Artifact()` call should stream the core dump over the service back to the caller.
```

The component that is specified should correspond to the software process that

> implements the BGP protocol in the system.

Reviewer: I think it would be very cool to augment this process with a few more details, which would clarify the interopretations between healthz service and telemetry paths.

1. shall we add here a step after the core dump is written that goes like: 
    > A Healthz event MUST be generated stating the new health state is UNHEALTHY. It is left for the implementor to decide which artifacts are generated during this automatic event generation following the rule that those automatic event generations SHOULD NOT affect the performance of a system. If the triage of this failure requires substantial computational resource, `Healthz.Check()` must be used by the client.
2. Clarify, that `Healthz.Get()` might not list (all?) the artifacts, if collection of those artifacts was considered to be resource intensive task. In such case, the `Check()` RPC should be used to populate the list of available artifacts.
3. Add a note, that telemetry paths are updated automatically, without calling any healthz RPCs denoting the changed health state

Reviewer: Hi @robshakir 
can you ptal at this comment? The rest LGTM (with a small term glitch [here](https://github.com/openconfig/gnoi/pull/106#discussion_r1012733954))

Author: I think this makes sense -- from a "safe" implementation perspective. Just looking at the place to add this to the doc.

Author: Added a paragraph that basically says exactly what you say further up in the document since I think it is important for the designers of clients too.

#### A chassis linecard goes unhealthy due to hardware failure.

##### Expected Action

Logs / dumps collected and the gNMI components are marked unhealthy.

A call to Healthz.Get should respond with the current state of the component
and provide feedback that there is an artifact available to be requested.

> A Healthz call should stream the core dump over the service back to the caller.

Reviewer: ```suggestion
A `Healthz.Artifact()` call should stream the core dump over the service back to the caller.
```

> A Healthz call should stream the logs over the service back to the caller.

Reviewer: ```suggestion
A `Healthz.Artifact()` call should stream the logs over the service back to the caller.
```

the component and no other action is necessary

The component reported on via telemetry should correspond to the device's chassis
and be of type
[`CHASSIS`](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#ident-chassis).

#### A user would like to stream console debugging or other shell output for collection

For a component healthz could be used to provide an I/O dump of a specific
debugging command redirected as a byte stream.  It could also be output like a
top or show interfaces where the data is periodically updated (each "frame"
would be sent over the stream as a separate message).  This would allow for
show tech like information to be collected and sent to vendors in a
programmatic secure way but not completely reinventing the wheel.

To achieve this use case, extensions to the existing specification are required.

#### Collect a snapshot of a particular component.

For components which can provide a "snapshot" of state the healthz artifact
endpoint can be provided to take a snapshot of state (ideally serialized as a
proto or other typed data structure) and provided to the caller. 

Examples:

>  * Get optical data from i2c bus on optical systems
>  * Get system database state for the core pub/sub of a device

Reviewer: ```suggestion
* Get optical data from i2c bus on optical systems
* Get system database state for the core pub/sub of a device
```

#### Exporting internal or specific debug logs

For a component in the system the developers could provide a debug log for that
specific component or other trace / dump information.  The `Check()` and
`Artifact()` APIs could be used in these cases to create the setup and then
download the specific data in a secure, reliable way (rather than a general file
scp). If the device can "automatically" generate these specific
data the `Get` API can also be used to check for existence and to download the

> artifacts. Note that to support this use case, asdditional gNMI paths are likely required to

Reviewer: ```suggestion
artifacts. Note that to support this use case, additional gNMI paths are likely required to
```

The ability to also include "implementation specific" details about the health
of the system or component is very standard for network devices in the form for
"show tech", "diag" or "debug" commands. These are very useful for getting
diagnostics about the system state to the implementers. Healthz exposes these
interfaces as queryable endpoints to allow operators to validate "health" on
components and if "unhealthy" gather implementor specific data to help triage
or reproduce issues.

## Design Background

`Healthz` provides a means by which a user may initiate checks of health
(through the `Check` RPC) or a system may report the results of a check that it
has initiated of its own accord.

Following a health check occurring, a caller uses the `List` or `Get` RPC to
discover the health "events" (expressed as a `ComponentStatus` message) that
are associated with a component and its corresponding subcomponents. Each event
reflects a collection of data that is required to debug or further root cause
the fault that occurs with an entity in the system.

The `Artifact` RPC is used to retrieve specific artifacts that are listed by

> the target system in the `List` or `Get` RPC.  Once retrieved an event, which

Reviewer: NIT: is this extra space needed? 
```suggestion
the target system in the `List` or `Get` RPC. Once retrieved an event, which
```

getting the "healthz" status of a component.  As discussed in the above issue,
due to large data volume or to an interactive debug output there is a
requirement that it support the ability to stream a status from the device.

To ensure that this can be done using a secure channel with the relevant ACL
and encryption, this streaming is implemented within the `Healthz` service.
This approach further allows the concept of a historical set of collected
artifacts to be reported by the device.

### Collection of Artifacts

As described above, by default health checks that are performed automatically
by the device SHOULD NOT be performance impacting. The `Check` RPC is used to
ensure the creation of artifacts that are expensive to collect.

Thus -- a system relying on `healthz` for reporting of device status, needs
to consider the following workflow.

1. A component of the system becomes unhealthy. The device MUST generate
   a health state via telemetry that indicates that the `healthz/state/status`
   of the component has transitioned to `UNHEALTHY`. It is left for the device
   implementor to determine which artifacts are generated automatically by the

>    system immediately. The artifact collection SHOULD NOt impact the performance

Reviewer: ```suggestion
   system immediately. The artifact collection SHOULD NOT impact the performance
```

   of the system.

> 2. The responding client system observes the transition to `UNHEALTHY` and can

Reviewer 2: this is an interesting idea
We would need to add an attribute which could signal there is "additional" work to be done to generate the artifacts 

It is easy to say "show tech" at the chassis is expensive but really we should never be getting "chassis" level checks we should really be only getting the needed artifacts for specific events

You do raise a reasonable use case though - let's say a system database reports an error - instead of pausing the system which could be service impacting to dumb the db- you could mark the event get the non service impacting data then leave a mark on the event that says - if you want more you can call check(event id) and it will go generate the other data and link to id which could be done after the device is drained or otherwise safe to get more data... That seems reasonable not sure it is "required" as we don't have a use case exactly of that

Author: Agreed. I've added the `event_id` to the `CheckRequest` to indicate that artifacts should be collected for a previous event, and made the clarifications that we discussed.

@hellt  PTAL at this too.

Reviewer: @robshakir where did you add the text/field about it? I can't seem to find new commits since [92516ec](https://github.com/openconfig/gnoi/pull/106/commits/92516ecaa5bea9611cda4c21d6a7f3166b66dd31)

Author: I'm sorry -- this was my fault. I'd not spotted that the push needed a merge. PTAL at https://github.com/openconfig/gnoi/pull/106/commits/8cd01e1000a03f86f4dbfec6e9fc40a0c0700d6c

considered "service impacting" and should have a clear security ACL restricting
their use during normal operations of the device.

The `CheckRequest` message includes an `event_id` field. When populated this
indicates that the `Check` should be performed for an event that has already
occurred within the system. The device should trigger artifact collection of
those artifacts that were not automatically collected. A `CheckRequest` for
a previous `event_id` MUST NOT overwrite previous artifacts that were collected
at the time of the event. The artifacts that are collected MUST be returned

> in the artifact list for the event when reported by the `Get` or `List` RPCs.

Reviewer: @robshakir I wonder how this is supposed to work, given that (I think) the health of a component can't be checked for the past events as the state of the component may already be HEALTHY or the artifacts collected won't contain the same state that were relevant at the time of a  healthy->unhealthy  transition.

I am inclined to not stall the PR though, as currently in our system we don't anticipate having expensive artifact collection processes which would require `Check` to work on a past `event_id`...

Reviewer: synced up offline
copying for historical purposes

> you call Check and it gathers some artifact that is based on the current state. This might include historical logs in the case of dumping some database, or retrieving some coredump from the secondary control plane.
But there's no guarantee that it captured the state at the time. For that, we need the artifacts that are automatically created to be complete.
